src/goto-programs/goto_contractor.cpp

```diff
@@ -171,37 +178,77 @@ void goto_contractort::contractor()
   log_status("{}", oss.str());
 }
 
-ibex::CmpOp goto_contractort::get_complement(ibex::CmpOp op)
+ibex::Ctc *goto_contractort::create_contractor_from_expr2t(const expr2tc &expr)
 {
-  switch(op)
+  ibex::Ctc *contractor = nullptr;
+  auto base_object = get_base_object(expr);
+
+  if(
+    (is_comp_expr(base_object) && !is_notequal2t(base_object)) ||
+    is_arith_expr(base_object))
+    contractor =
+      new ibex::CtcFwdBwd(*create_constraint_from_expr2t(base_object));
+  else
   {
-  case ibex::GEQ:
-    return ibex::LT;
-  case ibex::GT:
-    return ibex::LEQ;
-  case ibex::LEQ:
-    return ibex::GT;
-  case ibex::LT:
-    return ibex::GEQ;
-  default:
-    log_status("cant process equal");
-    break;
+    std::shared_ptr<logic_2ops> logic_op;
+    logic_op = std::dynamic_pointer_cast<logic_2ops>(base_object);
+
+    switch(base_object->expr_id)
+    {
+    case expr2t::expr_ids::and_id:
+      contractor = new ibex::CtcCompo(
+        *create_contractor_from_expr2t(logic_op->side_1),
+        *create_contractor_from_expr2t(logic_op->side_2));
+      break;
+    case expr2t::expr_ids::or_id:
+    {
+      auto side1 = create_contractor_from_expr2t(logic_op->side_1);
+      auto side2 = create_contractor_from_expr2t(logic_op->side_2);
+      contractor = new ibex::CtcUnion(*side1, *side2);
+      break;
+    }
+    case expr2t::expr_ids::notequal_id:
+    {
+      std::shared_ptr<relation_data> rel;
+      rel = std::dynamic_pointer_cast<relation_data>(base_object);
+      ibex::Function *f = create_function_from_expr2t(rel->side_1);
+      ibex::Function *g = create_function_from_expr2t(rel->side_2);
+      auto *side1 = new ibex::NumConstraint(*vars, (*f)(*vars) > (*g)(*vars));
+      auto *side2 = new ibex::NumConstraint(*vars, (*f)(*vars) < (*g)(*vars));
+      auto *c_side1 = new ibex::CtcFwdBwd(*side1);
+      auto *c_side2 = new ibex::CtcFwdBwd(*side2);
+      contractor = new ibex::CtcUnion(*c_side1, *c_side2);
+      break;
+    }
+    case expr2t::expr_ids::not_id:
+      //TODO: implement "not" flag  and process.
+      break;
+    default:
+      //unsupported. you shouldn't be here.
+      break;
+    }
   }
-  return ibex::GEQ;
+  return contractor;
 }
 
 ibex::NumConstraint *
-goto_contractort::create_constraint_from_expr2t(expr2tc expr)
+goto_contractort::create_constraint_from_expr2t(const expr2tc &expr)
 {
   ibex::NumConstraint *c;
+
   if(is_unsupported_operator(expr))
   {
     std::ostringstream oss;
     oss << get_expr_id(expr);
-    oss << "Expression is complex, skipping this assert.\n";
+    oss << " Expression is complex, skipping this assert.\n";
```

Comment:
Can you define the ``complex`` term?

Reply:
In Ibex, the expressions are structured more strictly than c. On the top level, we have logical operators such as && and ||. Then the comparison operators such as > and <. For example, we can have `(x > y || x < y)` and parse successfully. However, if the expression is valid in c `( x == y > 0 )`, it cannot be converted to ibex without reasoning. That's why we call them complex expressions.